### ISSUE TITLE
GPU accelerator (OpenCL / ML Drift) returns non-deterministic results for a single CONV_2D on Adreno 830

### ISSUE THREAD

**Issue by g-moschetti** [google-ai-edge/LiteRT#10192]
# GPU accelerator (OpenCL / ML Drift) returns non-deterministic results for a single CONV_2D on Adreno 830

## Summary

A model consisting of one `CONV_2D` (32 → 64 channels, 2×2 kernel, stride 2, no bias, random weights, fp32 NHWC I/O, input `[1, 336, 256, 32]`) produces **different outputs on consecutive invocations with the same input** when compiled for the GPU accelerator on a Snapdragon 8 Elite (Adreno 830). Each corrupted block is exactly one output row × 64 columns × 16 channels, i.e. one work-group tile, and the affected rows change from run to run. The CPU accelerator is bit-exact, and the GPU output matches the CPU except for the corrupted tiles.

Forcing `precision = 2` (fp32) **or** `buffer_storage_type = 1` (buffers instead of textures) in `gpu_options` makes the model deterministic. `enable_infinite_float_capping` and `hint_waiting_for_completion` do not. The fault is shape-dependent: other channel counts / resolutions of the same op are deterministic, and a 3×3 stride-2 convolution at the same shape is deterministic, which points at one kernel variant of the fp16 texture convolution path.

## Environment

| | |
|---|---|
| Device | Samsung Galaxy S25, Snapdragon 8 Elite `SM8750`, Adreno 830 |
| OS | Android 16, security patch 2026-08-05, `samsung/pa1qxeea/pa1q:16/BP4A.251205.006/S931BXXSCCZH1_OXMCCZH1` |
| GPU driver | `Adreno (TM) 830, OpenGL ES 3.2 V@0800.64.7 (GIT@f61dec9117, I68b0a64d7e, 1774510019) (Date:03/26/26)`, `/vendor/lib64/libOpenCL.so` |
| LiteRT | **v2.1.6** (commit `1461b6b2d`), prebuilt `litert/prebuilt/android_arm64/libLiteRt.so` (md5 `7d5f664b`) and `libLiteRtClGlAccelerator.so` (md5 `1a0a4c54`) from the repository, used through the C API |
| Also reproduced with | **v2.2.0** (commit `145c7523f`), prebuilt `libLiteRt.so` (md5 `489bd7d6`) and `libLiteRtClGlAccelerator.so` (md5 `81db30cb`), same harness binary |
| Model converter | `litert-torch` 0.9.2, `torch` 2.12.1 (`make_model.py`, attached) |

## Reproducer

Attached: `conv32_64_k2s2_336x256_fp32.tflite` (34 KB, one `CONV_2D`, fp32 in/out) and `make_model.py` to regenerate it:

```python
net = torch.nn.Sequential(torch.nn.Conv2d(32, 64, kernel_size=2, stride=2, bias=False)).eval()
nhwc = litert_torch.to_channel_last_io(net, args=[0], outputs=[0])
litert_torch.convert(nhwc, (torch.randn(1, 336, 256, 32),)).export("conv32_64_k2s2_336x256_fp32.tflite")
```

A second file, `conv32_64_k2s2_336x256_fp16.tflite` (19 KB, fp16 I/O with CAST ops and fp16 weights), behaves the same. Weights are irrelevant: any seed reproduces.

Harness (`probe.cpp`, attached, plain C API): create an environment, compile the model with `kLiteRtHwAcceleratorGpu`, fill the input once with a fixed pseudo-random pattern, call `LiteRtRunCompiledModel` N times, lock the host-memory output buffer after each run and compare it **bit for bit** with the first run. Any equivalent loop shows the same thing.

```
adb push litert_probe libLiteRt.so libLiteRtClGlAccelerator.so conv32_64_k2s2_336x256_fp32.tflite /data/local/tmp/p/
adb shell "cd /data/local/tmp/p && LD_LIBRARY_PATH=.:/vendor/lib64:/vendor/lib64/egl ./litert_probe conv32_64_k2s2_336x256_fp32.tflite --runs 100 --ref-cpu"
```

## Observed

Three launches of 100 runs each, fp32 model, default options:

```
L1: UNSTABLE in 99/99 runs, worst 3072 elements differ, max |diff| 1.43   vs CPU: 2013/1376256 elements differ
L2: UNSTABLE in 16/99 runs, worst 1024 elements differ
L3: UNSTABLE in 99/99 runs, worst 5120 elements differ
```

Same binary and model with the **v2.2.0** prebuilts, three launches of 100 runs:

```
L1: UNSTABLE in 11/99 runs, worst 2047 elements differ, max |diff| 1.83   vs CPU: 0 elements differ
L2: UNSTABLE in 98/99 runs, worst 3072 elements differ, max |diff| 1.65   vs CPU: 1004/1376256 elements differ (row 114)
L3: UNSTABLE in  9/99 runs, worst 2048 elements differ, max |diff| 1.83   vs CPU: 0 elements differ
```

With v2.2.0, `precision = 2` and `buffer_storage_type = 1` are again deterministic over 100 runs and bit-exact with the CPU.

Every differing block is one row of the `[1,168,128,64]` output, columns 64..127, channels `16k .. 16k+15`. Example from one launch (row: differing elements): `82:1024 110:1024 118:1024 149:1024`. The affected rows change between runs and between launches; the frequency per launch varies from a few percent to 100 %. Corrupted values are finite, no NaN/Inf. The first run of a launch can already differ from the CPU reference.

With `gpu_options` (opaque options payload, TOML):

| `gpu_options` | Result over 100 runs |
|---|---|
| *(none)* | unstable |
| `precision = 2` | deterministic, bit-exact with CPU |
| `buffer_storage_type = 1` | deterministic |
| `enable_infinite_float_capping = true` | unstable |
| `hint_waiting_for_completion = true` | unstable |

Same op, other shapes (random weights, fp16 I/O variant, 40–100 runs each):

| Conv2D 2×2 stride 2 | Input HxW | Result |
|---|---|---|
| **32 → 64** | **336×256** | **unstable** |
| **16 → 32** | **352×240** | **unstable** |
| 16 → 32 | 672×512 | deterministic |
| 32 → 64 | 176×120 | deterministic |
| 64 → 128 | 168×128 | deterministic |
| 64 → 128 | 88×60 | deterministic |
| 128 → 256 | 84×64 | deterministic |
| 128 → 256 | 44×30 | deterministic |
| 32 → 64, **3×3 stride 1** | 168×128 | deterministic |
| 32 → 64, **3×3 stride 2, pad 1** | 336×256 | deterministic |

## Expected

Identical output on every invocation for identical input, matching the CPU accelerator within fp16 tolerance.

## Related

- #8065 (LiteRT.js / WebGPU, Adreno 8xx) was root-caused by its reporter to an Adreno 830 texture-cache invalidation defect when a texture sampled by one dispatch is later overwritten by another, with buffer storage as the LiteRT-side mitigation. This report is very likely the same driver defect reached through the native OpenCL accelerator: same

 GPU, textures implicated (`buffer_storage_type = 1` cures it), no dependence on weights or input. Differences: here the corruption is intermittent per run and per launch rather than "every run after the first", it is confined to single work-group tiles, and the first run can already be affected.

## Workarounds

- `buffer_storage_type = 1` or `precision = 2` in `gpu_options`.
- Expressing the op as a 3×3 stride-2 pad-1 convolution with a zero first row and column of taps (numerically identical to the 2×2 stride-2 convolution) selects a different kernel and is deterministic.

## Attached files

[Conv2d_issue.zip](https://github.com/user-attachments/files/32563162/Conv2d_issue.zip)

**Comment by shivvaam0001** [comment]
Hi @g-moschetti,
Thank you for reaching out and bringing this issue to our attention.


It seems like this is caused by an Adreno 8xx driver texture cache invalidation defect in specific fp16 texture convolution paths, which is why forcing `buffer_storage_type = 1` or `precision = 2` bypasses it.

To help us investigate further, could you please clarify:

- Does this reproduce on other Adreno 8xx chips (e.g., Snapdragon 8 Gen 3), or is it strictly Adreno 830?
- What is the performance impact when using buffer_storage_type = 1 on your workload?

Thanks!

**Comment by g-moschetti** [comment]
Hello, I do not have another Adreno 8xx platform to test it out, unfortunately only 830.

The buffer_storage_type=1 impact was measured at around +30%.

For the moment the workaround I'm using is to avoid conv2x2 stride 2 with minimal impact on the performance.
